Improve tcp notifier performance

SeanTAllen opened the pull request:
Sending via TCP connections with writev is incredibly slow in
the default scenario. Pony's `error` keyword while much faster
than exceptions in Java is still relatively slow. By using,
error in the default case for `sentv` then any user code that
uses `TCPConnection.writev` which calls `sentv` will be very
slow. You'd have to know that you want to override `writev`
in order to get better performance.

The simple default here is to return `data` unchanged from
`sentv` rather than throwing the error.

This does have a negative impact on the user API in that
you can no longer easily use your custom handling in `sent`
from `sentv` without having to do anything.

After discussion during a Pony Sync meeting, it was decided
that this change was the correct thing to undertake.

Sylvan has expressed interest in adding on to this commit
with some changes to how SSL with operate with this change.
As such, this change works and compiles but perhaps shouldn't
be considered complete as is.

SeanTAllen commented:
DO NOT MERGE YET.

@sylvanc please make your changes that you indicated you wanted to do for SSLConnection as part of this change and then we can merge.

SeanTAllen commented:
N.B. this is a breaking API change in case it isnt obvious from the commit message (I didn't explicitly state it).

SeanTAllen commented:
@jemc I should probably add a CHANGELOG entry yes? thoughts on what that should be?

jemc commented:
I think a reasonable changelog entry would be: "Changed expected behavior of TCPNotify.sent and sentv methods".

jemc commented:
Also, I noticed that you removed the `_TestTCPWritev` test.  We'll want to replace this with a test that proves the new behavior before merging.

SeanTAllen commented:
Yeah the previous test no longer made sense. 

What is a good test for `sent` and `sentv`, 

"hi" enters and "bye" leaves?

"hi" enters and "hi there" leaves?

jemc commented:
The most basic test I can think of probably wouldn't be *that* different, just testing that `sent` is called for `write`, that `sentv` is called for `writev`, and that in both cases the notifier can transform the bytes to something different.

SeanTAllen commented:
@sylvanc do you want to add your changes in to this still to fix SSL or should I add the test @jemc requested and merge?

SeanTAllen commented:
this needs a CHANGELOG entry.

SeanTAllen commented:
@sylvanc please put your SSL changes that are needed in here and I will implement.

SeanTAllen commented:
@jemc test added. take a look.

I think I found exactly what the SSL issue is and we can merge this without it and I am opening an issue for the SSL issue.

SeanTAllen commented:
@jemc also I fucked up and merged from master instead of rebase so... we should probably just merge this in without squashing etc. I'm not sure what will happen if you squash in the merge from master. blarg. sorry.

SeanTAllen commented:
Also, I think this deserves a release (its a big perf bump) and because it breaks the public API, we should call it 0.4.0

Thoughts @ponylang/committer ?

SeanTAllen commented:
Client connect failed in the writev test on Appveyor. We seem to have a lot of issues with network tests on Appveyor. Not sure why they only occassionaly fail though. Maybe sometimes it take a while to get a connection? I'm wondering, should we try upping timeouts to see if it goes away @jemc? Its a fairly awful idea but if that is what it takes not to get false positives...

SeanTAllen commented:
all the appveyor writev tests failed again on connect. not sure what the issue with that is. hmmm...

could it be a timing issue with appveyor and having multiple network tests running?

jemc commented:
@SeanTAllen Thing is, it doesn't seem like it's a timeout issue, at least at first glance - the action is failing, not timing out:

```
**** FAILED: net/TCP.writev
Action failed: client connect
Complete(false) called
```

It also looks like it failed consistently across all four appveyor tests...

SeanTAllen commented:
@jemc i think i know what a timing issue might be.

SeanTAllen commented:
@jemc terrifyingly i think i fixed it. i changed the TCPConnection that _TCPTest opens from specifically using `TCPConnection.ip4` to `TCPConnection` and now everything is passing.

SeanTAllen commented:
@jemc if everything passes, I'm going to remove a couple of changes, put the ip4 change back in and `push -f` then if it passes again, we can merge. Sound good?

jemc commented:
:+1:

SeanTAllen commented:
forced pushed @jemc. fingers crossed.

jemc commented:
:tada:

SeanTAllen commented:
FINALLY
